controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+			log.Info("updating processes of existing formation", "processes", ef.Processes)
+			f.Processes = ef.Processes
+		}
+	}
+	return f, nil
+}
+
+func (s *Scheduler) updateFormation(f *ct.Formation) (*Formation, error) {
+	ef, err := utils.ExpandFormation(s, f)
+	if err != nil {
+		return nil, err
+	}
+	return s.changeFormation(ef)
+}
+
+func (s *Scheduler) startJob(req *JobRequest) (err error) {
+	log := fnLogger("job.type", req.Type)
+	log.Info("starting job", "job.restarts", req.restarts, "request.attempts", req.attempts)
+	// Copy on Write
+	s.handleJobStop(req.Job)
+	newReq := req.Clone()
+	newReq.HostID = ""
+	newReq.JobID = random.UUID()
+	newReq.state = JobStateRequesting
+	defer func() {
+		if err == nil || newReq.attempts < maxJobAttempts {
+			s.jobs[newReq.JobID] = newReq.Job
+			if err != nil {
+				log.Error("error starting job, trying again", "err", err)
+				newReq.attempts++
+				time.AfterFunc(jobAttemptInterval, func() {
+					s.jobRequests <- newReq
+				})
+			}
+		}
+	}()
+
+	log.Info("determining best host for job")
+	host, err := s.findBestHost(newReq.Formation, newReq.Type, newReq.HostID)
+	if err != nil {
+		log.Error("error determining best host for job", "err", err)
+		return err
+	}
+	hostID := host.ID()
+	newReq.HostID = hostID
+
+	config := jobConfig(newReq, hostID)
+	newReq.JobID = config.ID
+
+	// Provision a data volume on the host if needed.
+	if newReq.needsVolume() {
+		log.Info("provisioning volume")
+		if err := utils.ProvisionVolume(host, config); err != nil {
+			log.Error("error provisioning volume", "err", err)
+			return err
+		}
+	}
+
+	log.Info("requesting host to add job", "host.id", hostID, "job.id", config.ID)
+	if err := host.AddJob(config); err != nil {
+		log.Error("error requesting host to add job", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (s *Scheduler) stopJob(req *JobRequest) (err error) {
+	log := fnLogger("req.host.id", req.HostID, "req.job.id", req.JobID, "job.type", req.Type)
+	log.Info("stopping job")
+	defer func() {
+		if err != nil {
+			log.Error("error stopping job", "err", err)
+		}
+	}()
+
+	var job *Job
+	if req.JobID == "" {
+		formationKey := utils.FormationKey{AppID: req.AppID, ReleaseID: req.ReleaseID}
+		typJobs := s.jobs.GetStoppableJobs(formationKey, req.Type)
+		if len(typJobs) == 0 {
+			e := fmt.Sprintf("no %s jobs running", req.Type)
+			log.Error(e)
+			return errors.New(e)
+		}
+		job = typJobs[0]
+		for _, j := range typJobs {
+			if j.startedAt.After(job.startedAt) {
+				job = j
+			}
+		}
+		log.Info("selected job for termination", "job.id", job.JobID, "job.host.id", job.HostID)
+	} else {
+		var ok bool
+		job, ok = s.jobs[req.JobID]
+		if !ok {
+			e := "unknown job"
+			log.Error(e)
+			return errors.New(e)
+		}
+	}
+	job = s.handleJobStop(job)
+	if job.HostID != "" {
+		log.Info("getting host client")
+		host, err := s.Host(job.HostID)
+		if err != nil {
+			log.Error("error getting host client", "err", err)
+			s.unfollowHost(job.HostID)
```

Comment:
It won't be once we implement SyncHosts and correctly handle intermittent host disconnecting. I'll add a TODO.